tests/ga/test_update.py

```diff
@@ -1479,6 +1486,118 @@ def test_telemetry_heartbeat_creates_event(self, patch_add_event, patch_info, *_
         self.assertTrue(any(call_args[0] == "[HEARTBEAT] Agent {0} is running as the goal state agent {1}"
                             for call_args in patch_info.call_args), "The heartbeat was not written to the agent's log")
 
+    @contextlib.contextmanager
+    def _get_update_handler(self, iterations=1, test_data=DATA_FILE):
+        """
+        This function returns a mocked version of the UpdateHandler object to be used for testing. It will only run the
+        main loop [iterations] no of times.
+        To reuse the same object, be sure to reset the iterations by using the update_handler.set_iterations() function.
+        :param iterations: No of times the UpdateHandler.run() method should run.
+        :return: Mocked object of UpdateHandler() class and object of the MockWireProtocol().
+        """
+
+        def _set_iterations(iterations):
+            # This will reset the current iteration and the max iterations to run for this test object.
+            update_handler._cur_iteration = 0
+            update_handler._iterations = iterations
+
+        def check_running(*args, **kwargs):
+            # This method will determine if the current UpdateHandler object is supposed to run or not.
+            if update_handler._cur_iteration < update_handler._iterations:
+                update_handler._cur_iteration += 1
+                return True
+            return False
+
+        with mock_wire_protocol(test_data) as protocol:
+            protocol_util = MagicMock()
+            protocol_util.get_protocol = Mock(return_value=protocol)
+            with patch("azurelinuxagent.ga.update.get_protocol_util", return_value=protocol_util):
+                with patch("azurelinuxagent.common.conf.get_autoupdate_enabled", return_value=False):
+                    update_handler = get_update_handler()
+                    # Setup internal state for the object required for testing
+                    update_handler._cur_iteration = 0
+                    update_handler._iterations = 0
+                    update_handler.set_iterations = lambda i: _set_iterations(i)
+                    type(update_handler).running = PropertyMock(side_effect=check_running)
+                    with patch("time.sleep", side_effect=lambda _: mock_sleep(0.001)):
+                        with patch('sys.exit'):
+                            # Setup the initial number of iterations
+                            update_handler.set_iterations(iterations)
+                            try:
+                                yield update_handler, protocol
+                            finally:
+                                # Since PropertyMock requires us to mock the type(ClassName).property of the object,
+                                # reverting it back to keep the state of the test clean
+                                type(update_handler).running = True
```

Comment:
Couldn't you also reset the iterations counter here?

Reply:
This block (`finally`) is only hit when we exit the block where we initialize the update handler. The whole idea was to re-use the same object without creating a new mock_update_handler for every small run. Does that answer your question? I can explain more offline if needed

Reply:
Got it! Thanks.